lib/core/networking/protocol/protocol.dart

```diff
@@ -0,0 +1,46 @@
+import 'dart:convert';
+
+import 'authentication.dart';
+
+typedef Deserialize<T> = T Function(Map<String, dynamic>);
+
+class ProtocolDelegate {
+  /// A dictionary of all protocol elements and their coresponding deserializers
+  //? This should not be final because of how testing is implemented.
+  static Map<String, Deserialize> elements = <String, Deserialize>{
+    //* Authentication
+    'AuthRequest': (Map<String, dynamic> json) => AuthRequest.fromJson(json),
+    'AuthResponse': (Map<String, dynamic> json) => AuthResponse.fromJson(json),
+  };
+
+  static String serialize(dynamic message) {
+    // Get className
+    final String type = message.runtimeType.toString();
+    final String data = jsonEncode(message);
```

Comment:
if you call `message.toJson()` you could get an error on unsupported data types or does `jsonEncode` already handles this? And why don't you check if tyoe is in `elements`?

Reply:
All protocol messages (elements) are `@JsonSerializable`->this will always work. We dont check in elements as we:
1. trust ourselfs
2. this would just raise an error -> why throw a seperate one? The current one would be way more informative.